toplevel/toploop.ml

```diff
@@ -206,21 +218,42 @@ let loop ppf =
   run_hooks After_setup;
   load_ocamlinit ppf;
   while true do
-    let snap = Btype.snapshot () in
+    let snap = ref (Btype.snapshot ()) in
     try
       Lexing.flush_input lb;
       (* Reset the phrase buffer when we flush the lexing buffer. *)
       Buffer.reset phrase_buffer;
       Location.reset();
-      Warnings.reset_fatal ();
       first_line := true;
-      let phr = try !parse_toplevel_phrase lb with Exit -> raise PPerror in
-      let phr = preprocess_phrase ppf phr  in
-      Env.reset_cache_toplevel ();
-      ignore(execute_phrase true ppf phr)
+      let rec get_phrases phrs =
+        match !parse_toplevel_phrase lb with
+        | phr ->
+          if is_blank lb then
+            List.rev (phr :: phrs)
+          else
+            get_phrases (phr :: phrs)
+        | exception Exit -> raise PPerror
+        | exception e -> Location.report_exception ppf e; []
+      in
+      let phrs = get_phrases [] in
+      let saved_report_printer = !Location.report_printer in
+      let nonfirst_report_printer = fun () -> Location.batch_mode_printer in
+      let snap = ref (Btype.snapshot ()) in
```

Comment:
Unless I am missing something, this line introduces a bug because it shadows the `snap` reference on line 221 which is the one seen by the exception handlers below.

Reply:
Good catch, thanks.